README.md
# 📈 PHP Zabbix Graph

[![Latest Version on Packagist](https://img.shields.io/packagist/v/casperboone/zabbix-graph.svg?style=flat-square)](https://packagist.org/packages/casperboone/zabbix-graph)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![StyleCI](https://styleci.io/repos/86865582/shield)](https://styleci.io/repos/86865582)
[![Build Status](https://img.shields.io/travis/casperboone/zabbix-graph/master.svg?style=flat-square)](https://travis-ci.org/casperboone/zabbix-graph)
[![SensioLabsInsight](https://img.shields.io/sensiolabs/i/09228096-56d8-4618-93a0-92bacc86a0c9.svg?style=flat-square)](https://insight.sensiolabs.com/projects/09228096-56d8-4618-93a0-92bacc86a0c9)
[![Quality Score](https://img.shields.io/scrutinizer/g/casperboone/zabbix-graph.svg?style=flat-square)](https://scrutinizer-ci.com/g/casperboone/zabbix-graph)
[![Code Coverage](https://img.shields.io/scrutinizer/coverage/g/casperboone/zabbix-graph/master.svg?style=flat-square)](https://scrutinizer-ci.com/g/casperboone/zabbix-graph/?branch=master)

Get a graph from Zabbix to display on a webpage or save to a file. If you are using Laravel, then please check out [this repository](https://github.com/casperboone/laravel-zabbix-graph). 

## Installation
You can install the package via composer:

``` bash
composer require casperboone/zabbix-graph
```

Require Composer's autoload (probably already done):
```php
require __DIR__.'/../vendor/autoload.php';
```


## Usage
### Basic Usage
You can create an instance of `CasperBoone\ZabbixGraph` and pass the full URL to your Zabbix installation, the username and the password through the constructor. On this instance you can get a graph from Zabbix by calling `->find($graphId)`. Graph IDs can be found in the URL of the Zabbix UI of a certain graph.

Example:
```php
$zabbixGraph = new CasperBoone\ZabbixGraph('http://my-zabbix.com', 'username', 'passsword');

$zabbixGraph->width(500)
    ->height(300)
    ->find(54);
```

The output of find is a binary image that can be saved to a file or converted to an HTTP response.

### Available Methods
The following methods are available to adjust the parameters of the graph:

| Method                          | Description                                                |
| ------------------------------- | ---------------------------------------------------------- |
| `->width(int $width)`           | The width of the graph in pixels*                          |
| `->height(int $width)`          | The height of the graph in pixels*                         |
| `->startTime(DateTime $start)`  | The start date and time of the data displayed in the graph |
| `->endTime(DateTime $end)`      | The end date and time of the data displayed in the graph   |

_* The graph that Zabbix returns is usually slightly bigger because of added legends or labels_
### Old Zabbix versions
If you're using Zabbix 1.8 or older, then you need to set the last parameter of the constructor to `true`. 

Example:
```php
$zabbixGraph = new CasperBoone\ZabbixGraph('http://my-zabbix.com', 'username', 'passsword', true);
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

One of the integration tests requires a non public Zabbix installation. In order to execute this test, you have to create a `.env` file in the project root and copy the contents of `.env.example` into it. Complete the `.env` file with the host, a username/password and an available graph ID.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email mail@casperboone.nl instead of using the issue tracker.

## Credits

- [Casper Boone](https://github.com/casperboone)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `casperboone/zabbix-graph` will be documented in this file

## 1.0.0 - 2017-04-01
- Initial release


tests/TestCase.php
<?php

namespace CasperBoone\ZabbixGraph\Test;

use Dotenv\Dotenv;
use Dotenv\Exception\InvalidPathException;
use PHPUnit\Framework\TestCase as PHPUnitTestCase;

class TestCase extends PHPUnitTestCase
{
    public function loadenv($absentEnvMessage = null)
    {
        try {
            (new Dotenv(__DIR__.'/../'))->load();
        } catch (InvalidPathException $exception) {
            $this->markTestSkipped($absentEnvMessage ?? 'Env file not present');

            return;
        }
    }
}


tests/Unit/FakeZabbixGraph.php
<?php

namespace CasperBoone\ZabbixGraph\Test\Unit;

use Mockery;
use CasperBoone\ZabbixGraph\ZabbixGraph;

class FakeZabbixGraph extends ZabbixGraph
{
    /**
     * Create HTTP client for requesting the graph. The HTTP client should preserve cookies.
     *
     * @param  string $url Full url of Zabbix location
     * @return HttpClient
     */
    protected function createHttpClient($url)
    {
        $this->url = $url;

        return Mockery::spy(\GuzzleHttp\ClientInterface::class);
    }

    /**
     * @return \GuzzleHttp\Client
     */
    public function httpClient()
    {
        return $this->httpClient;
    }
}


tests/Unit/ZabbixGraphTest.php
<?php

namespace CasperBoone\ZabbixGraph\Test\Unit;

use Mockery;
use PHPUnit\Framework\TestCase;
use Psr\Http\Message\MessageInterface;
use Psr\Http\Message\ResponseInterface;

class ZabbixGraphTest extends TestCase
{
    /**
     * @var FakeZabbixGraph
     */
    private $zabbixGraph;

    protected function setUp()
    {
        parent::setUp();

        $this->zabbixGraph = new FakeZabbixGraph('http://example.com', 'john', 'welcome');

        $this->setUpMocks();
    }

    /** @test */
    public function it_can_login_to_zabbix()
    {
        $this->zabbixGraph->find(1);

        $this->assertRequestHasField('name', 'john', 'post');
        $this->assertRequestHasField('password', 'welcome', 'post');

        // Verify if path is correct
        $this->assertEquals('http://example.com', $this->zabbixGraph->url);
        $this->zabbixGraph->httpClient()->shouldHaveReceived('post')->with('index.php', Mockery::any());
    }

    /** @test */
    public function it_can_request_a_graph_with_correct_defaults()
    {
        $image = $this->zabbixGraph->find(123);

        $this->assertEquals('image', $image);
        $this->assertRequestHasField('graphid', 123);
        $this->assertRequestHasField('width', 400);
        $this->assertRequestHasField('height', 200);

        // Verify if path is correct
        $this->zabbixGraph->httpClient()->shouldHaveReceived('get')->with('chart2.php', Mockery::any());
    }

    /** @test */
    public function a_custom_width_can_be_set_and_is_included_in_the_request()
    {
        $this->zabbixGraph
            ->width(123)
            ->find(1);

        $this->assertRequestHasField('width', 123);
    }

    /** @test */
    public function a_custom_height_can_be_set_and_is_included_in_the_request()
    {
        $this->zabbixGraph
            ->height(123)
            ->find(1);

        $this->assertRequestHasField('height', 123);
    }

    /** @test */
    public function a_custom_start_time_can_be_set_and_is_included_in_the_request()
    {
        $this->zabbixGraph
            ->startTime((new \DateTime())->setTimestamp(1445758080))
            ->find(1);

        $this->assertRequestHasField('stime', 1445758080);
    }

    /** @test */
    public function a_custom_end_time_can_be_set_and_is_included_in_the_request()
    {
        $this->zabbixGraph
            ->startTime((new \DateTime())->setTimestamp(10))
            ->endTime((new \DateTime())->setTimestamp(100))
            ->find(1);

        $this->assertRequestHasField('period', 90);
    }

    private function assertRequestHasField($key, $value, $method = 'get')
    {
        $requestDataKey = [
            'get' => 'query',
            'post' => 'form_params',
        ][$method];

        $this->zabbixGraph->httpClient()->shouldHaveReceived($method)
            ->with(Mockery::any(), Mockery::on(function ($options) use ($requestDataKey, $key, $value) {
                $this->assertEquals($value, $options[$requestDataKey][$key]);

                return true;
            }));
    }

    private function setUpMocks()
    {
        $this->zabbixGraph->httpClient()->shouldReceive('get')
            ->andReturn($response = Mockery::mock(ResponseInterface::class));
        $response->shouldReceive('getBody')
            ->andReturn($message = Mockery::mock(MessageInterface::class));
        $message->shouldReceive('getContents')
            ->andReturn('image');
    }
}


src/ZabbixGraph.php
<?php

namespace CasperBoone\ZabbixGraph;

use DateTime;
use GuzzleHttp\Client as HttpClient;

class ZabbixGraph
{
    protected $httpClient;
    protected $oldZabbix;
    protected $username;
    protected $password;

    protected $width = 400;
    protected $height = 200;
    protected $startTime;
    protected $endTime;

    /**
     * Construct and initalize ZabbixGraph.
     *
     * @param  string  $url        Full url of Zabbix location
     * @param  string  $username   Zabbix username
     * @param  string  $password   Zabbix password
     * @param  bool    $oldZabbix  Set to true if using Zabbix 1.8 or older, by default set to false
     */
    public function __construct($url, $username, $password, $oldZabbix = false)
    {
        $this->httpClient = $this->createHttpClient($url);
        $this->oldZabbix = $oldZabbix;
        $this->username = $username;
        $this->password = $password;

        $this->startTime = (new DateTime())->modify('-1 hour');
        $this->endTime = new DateTime();
    }

    /**
     * Create HTTP client for requesting the graph. The HTTP client should preserve cookies.
     *
     * @param  string  $url  Full url of Zabbix location
     * @return HttpClient
     */
    protected function createHttpClient($url)
    {
        return new HttpClient([
            'base_uri' => $url,
            'cookies' => true,
        ]);
    }

    /**
     * Request graph from Zabbix and return a raw image. If an error
     * occurred Zabbix will output this as an image.
     *
     * @param  int  $graphId  ID of the graph to be requested
     * @return string
     */
    public function find($graphId)
    {
        $this->login();

        return $this->httpClient->get('chart2.php', [
            'query' => [
                'graphid' => $graphId,
                'width' => $this->width,
                'height' => $this->height,
                'stime' => $this->startTime->getTimestamp(),
                'period' => $this->endTime->getTimestamp() - $this->startTime->getTimestamp(),
            ],
        ])->getBody()->getContents();
    }

    /**
     * Login to Zabbix to acquire the needed session for requesting the graph.
     */
    protected function login()
    {
        $this->httpClient->post('index.php', [
            'form_params' => [
                'name' => $this->username,
                'password' => $this->password,
                'enter' => 'Sign in',
            ],
        ]);
    }

    /**
     * Specify width of the graph in pixels, by default 400.
     *
     * @param  int  $width  Width in pixels
     * @return $this
     */
    public function width($width)
    {
        $this->width = $width;

        return $this;
    }

    /**
     * Specify height of the graph in pixels, by default 400.
     *
     * @param  int  $height  Height in pixels
     * @return $this
     */
    public function height($height)
    {
        $this->height = $height;

        return $this;
    }

    /**
     * Specify start date and time of the data displayed in the graph.
     *
     * @param  DateTime  $startTime  Start date and time
     * @return $this
     */
    public function startTime(DateTime $startTime)
    {
        $this->startTime = $startTime;

        return $this;
    }

    /**
     * Specify end date and time of the data displayed in the graph.
     *
     * @param  DateTime  $endTime  End date and time
     * @return $this
     */
    public function endTime(DateTime $endTime)
    {
        $this->endTime = $endTime;

        return $this;
    }
}
